build.go

```
@@ -284,13 +285,26 @@ func Build(f BuildFunc, options ...Option) {
 		}
 
 		var launch struct {
-			Processes []Process  `toml:"processes"`
-			Slices    []Slice    `toml:"slices"`
-			Labels    []label    `toml:"labels"`
-			BOM       []BOMEntry `toml:"bom"`
+			Processes       []Process       `toml:"processes,omitempty"`
+			DirectProcesses []DirectProcess `toml:"processes,omitempty"`
+			Slices          []Slice         `toml:"slices"`
+			Labels          []label         `toml:"labels"`
+			BOM             []BOMEntry      `toml:"bom"`
+		}
+
+		if apiVersion.LessThan(apiV09) {
+			launch.Processes = result.Launch.Processes
+		} else {
+			launch.DirectProcesses = []DirectProcess{}
+			if result.Launch.Processes != nil {
+				for _, process := range result.Launch.Processes {
+					launch.DirectProcesses = append(launch.DirectProcesses, process.ToDirectProcess())
```

I'm trying to decide what my stance is on converting the any non-direct processes into direct ones, versus. issuing an error when the process isn't well-formed for the API version.

On one hand, doing the conversion you have here makes the buildpack-author experience simpler by abstracting away the specifics. It also makes the upgrade path to API v0.9 easier. I think I prefer this.

On the other hand, it could be confusing to a buildpack author that a process is being converted under the hood. It might be more straightforward to just enforce the correct process structure.

@paketo-buildpacks/tooling-maintainers any opinions?

---

My understanding is that API v0.9 pushes the decision to use a shell back to the individual builpack instead of deciding it centrally in the `lifecycle` (via indirect process). If a buildpack maintainer consciously decides to update to v0.9 maybe it's better to consciously decide on what and how processes should be started, instead of again deciding it centrally in `packit`. On the other hand side, `packit` is more high-level library from which uses might expect that level of convenience of auto-conversion.

This conversion could also be added in a separate PR on demand.

---

This is a great conversation. I can see the pros and cons of both approaches - packit being dumb and relying on the buildpack author to do the right thing, or packit trying to be opinionated and save the buildpack author from extra work.

My perspective is that it packit should stay fairly dumb on this point. I think it would be really confusing to debug a situation where a buildpack claimed it was compatible with API `v0.9`, and provided indirect processes, and yet still magically worked because packit automatically converted the indirect processes to direct processes.

As a buildpack author, if I upgrade to API `v0.9` I would expect to make the necessary changes to my buildpack to ensure it is compatible with that updated API version. I would be surprised if it just worked because Packit intervened, and I'd be worried that packit might do the wrong thing with my command in some other situation.

I think it's better to fail fast and inform the user/buildpack author that the process type must be direct if the API is `v0.9` or higher.

It looks similar to the situation where we fail fast if a process is marked as `Default` for API `v0.5` or lower. In this situation, Packit could be more intelligent and try to remove the `Default` label - especially if there is only one process - but I think it's more useful to fail and tell the user that they're trying to do something that contravenes the spec.

---

Thanks both for weighing in, it is an interesting problem :). I think Rob's point that we have the previous experience of failing fast in the `v0.5` API default process example he provided is good enough reason to move forward without auto process conversion.
I totally see the value in doing the conversion, but it's probably more important to keep the packit experience consistent.

---

Thanks for the input and sorry for the delay. I'll try to get the PR ready this week.

---

No rush!